spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,223 @@
+
+The logic involved in collecting fees from users and then paying it out to the relevant relayers is encapsulated by a separate fee module and may vary between implementations. However, all fee modules must implement a uniform interface such that the ICS-4 handlers can correctly pay out fees to the right relayers, and so that relayers themselves can easily determine the fees they can expect for relaying a packet.
+
+### Fee Module Contract
+
+While the details may vary between fee modules, all Fee modules **must** ensure it does the following:
+
+- It must have in escrow the maximum fees that all outstanding packets may pay out (or it must have ability to mint required amount of tokens)
+- It must pay the receive fee for a packet to the forward relayer specified in `PayFee` callback
+- It must pay the ack fee for a packet to the reverse relayer specified in `PayFee` callback
+- It must pay the timeout fee for a packet to the timeout relayer specified in `PayTimeoutFee` callback
+- It must refund any remainder fees in escrow to the original fee payer(s) if applicable
+
+```typescript
+function PayFee(packet: Packet, forward_relayer: string, reverse_relayer: string) {
+    // pay the forward fee to the forward relayer address
+    // pay the reverse fee to the reverse relayer address
+    // refund extra tokens to original fee payer(s)
+}
+
+function PayTimeoutFee(packet: Packet, timeout_relayer: string) {
+    // pay the timeout fee to the timeout relayer address
+    // refund extra tokens to original fee payer(s)
+}
+```
+
+
+The fee module should also expose the following queries so that relayers may query their expected fee:
+
+```typescript
+// Gets the fee expected for submitting ReceivePacket msg for the given packet
+function GetReceiveFee(portID, channelID, sequence) Fee
+
+// Gets the fee expected for submitting AcknowledgePacket msg for the given packet
+function GetAckFee(portID, channelID, sequence) Fee
+
+// Gets the fee expected for submitting TimeoutPacket msg for the given packet
+function GetTimeoutFee(portID, channelID, sequence) Fee
+```
+
+Since different chains may have different representations for fungible tokens and this information is not being sent to other chains; this ICS does not specify a particular representation for the `Fee`. Each chain may choose its own representation, it is incumbent on relayers to interpret the Fee correctly.
+
+### Connection Negotiation
+
+The chains must agree to enable the incentivization feature during the connection handshake. This can be done by bumping the connection version.
```

Comment:
Wait, why at the connection level instead of at the channel level? Different channels for different applications - but on the same connection - might have totally different fee models.

Reply:
The connection is just negotiating the general fee feature. It is in effect asking if the counterparty chain will send the forward_relayer back in the acknowledgement. That's it. This is not a feature of the application, this is a feature of the chain; which is why we're negotiating on connection so we have a uniform way to negotiate and send/receive relayer information in ICS-4 handlers rather than having all IBC apps implement this feature themselves and encode core IBC feature support into their app-specific channel versions.
 
Since this proposal handles fee payment logic completely on a single chain, we are not negotiating anything specific about fee-handling logic. Both apps can implement completely different fee models using the interfaces in this spec. So long as each chain is propogating relayer information to each other correctly, they can operate their different fee models as much as they like. It does not need to be agreed upon even between apps on opposite ends of a channel. One side could do sender-incentivized packets while the other side uses an altruistic fund.